dao/src/main/java/com/epam/esm/dao/impl/CertificateDaoImpl.java

```diff
@@ -0,0 +1,102 @@
+package com.epam.esm.dao.impl;
+
+import com.epam.esm.dao.CertificateDao;
+import com.epam.esm.dao.specification.Specification;
+import com.epam.esm.entity.Certificate;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.jdbc.core.BeanPropertyRowMapper;
+import org.springframework.jdbc.core.JdbcTemplate;
+import org.springframework.stereotype.Component;
+
+import java.util.List;
+import java.util.Optional;
+
+@Component
+public class CertificateDaoImpl implements CertificateDao {
+    private final JdbcTemplate jdbcTemplate;
+
+    private final String FIND_BY_ID = """
+            SELECT id, name, description, price, duration, create_date, last_update_date FROM certificates WHERE id=?
+            """;
+    private final String SAVE = """
+            INSERT INTO certificates (name, description, price, duration)
+            VALUES(?,?,?,?) RETURNING id, name, description, price, duration, create_date, last_update_date
+            """;
+    private final String UPDATE = """
+            UPDATE certificates
+            SET
+            name=COALESCE(?, name),
+            description=COALESCE(?, description),
+            price=COALESCE(?, price),
+            duration=COALESCE(?, duration)
+            WHERE id=?
+            RETURNING id, name, description, price, duration, create_date, last_update_date
+            """;
+    private final String REMOVE_BY_ID = """
+            DELETE FROM certificates WHERE id=?
+            """;
+    private final String REMOVE_ALL_TAGS_BY_CERTIFICATE_ID = """
+            DELETE FROM tag_certificate_membership WHERE certificate_id=?
+            """;
+    private final String ADD_TAG_TO_CERTIFICATE = """
+            INSERT INTO tag_certificate_membership (tag_id, certificate_id) VALUES(?,?)
+            """;
+
+    @Autowired
+    public CertificateDaoImpl(JdbcTemplate jdbcTemplate) {
+        this.jdbcTemplate = jdbcTemplate;
+    }
+
+    @Override
+    public List<Certificate> findBySpecification(Specification specification, String sortParams) {
```

Comment:
It looks like that **sortParams** parameter is a sql query part that is formed in service layer. If true:

1. Service layer: How to know what sortParams are allowed?
2. Service layer: What if it will be decided that dao should work is NoSql or sent data to another controller through web?
3. Dao layer: How to prevent injecting undesired queries from Service layer?

Reply:
Fixed

Reply:
How?